File: Csharp/Creational/Singleton/Example/Singleton/Singleton.cs
namespace Progam;

public class Singleton
{
    private static Singleton? _instance;
    private static readonly object _lock = new object();
    
    private Singleton()
    {
        Console.WriteLine("Singleton instance created.");
    }
    public static Singleton Instance()
    {
        if (_instance == null)  // First check (without lock)
        {
            lock (_lock)  // Ensure thread safety
            {
                if (_instance == null)  // Second check (inside lock)
                {
                    _instance = new Singleton();
                }
            }
        }
        return _instance;
    }
    public void ShowMessage()
    {
        Console.WriteLine("Hello from Singleton!");
    }
    
}

File: Java/Creational/Singleton/Example1/Main/src/Singleton.java
public class Singleton {
    private static volatile Singleton instance;
    private Singleton() {
        System.out.println("Singleton instance created.");
    }

    public static Singleton getInstance() {
        if (instance == null) { // First check (without lock)
            synchronized (Singleton.class) { // Ensure thread safety
                if (instance == null) { // Second check (inside synchronized)
                    instance = new Singleton();
                }
            }
        }
        return instance;
    }

    public void showMessage() {
        System.out.println("Hello from Singleton!");
    }
}